### PR TITLE
Define example difficulty

### PR DESCRIPTION
**Pull request by sshleifer**
So that I can make a comment

### CODE REVIEW

File: fastai/imagito/sample_hardness.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+             model_params['bs'],
+             model_params['sample'])
+
+    l = Learner(data, m(c_out=10), path=model_dir)
+    l.load('final_classif')
+    return l, data
+
+def choose_n(t, n):
+    """Choose n random elements from tensor t"""
+    idx = torch.randperm(t.size(0))[:n]
```
Comment (sshleifer) [inline fastai/imagito/sample_hardness.py:27]: `torch.randperm(n)`?

File: fastai/imagito/sample_hardness.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+def do_filter(ll:LabelList, mask):
+    """
+    Filter images in LabelList where mask == 0
+    Mutates `ll`
+    """
+    global i
+    i = 0
+    def mask_filter(img):
+        global i
```
Comment (sshleifer) [inline fastai/imagito/sample_hardness.py:48]: does this need global?

File: fastai/imagito/sample_hardness.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        keep = bool(mask[i])
```
Comment (sshleifer) [inline fastai/imagito/sample_hardness.py]: dont think you need cast

File: fastai/imagito/sample_hardness.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        i += 1
+        return keep
+
+    ll.x.filter_by_func(mask_filter)
+    i = 0
+    ll.y.filter_by_func(mask_filter)
+    return ll
+
+def proxy_dist(sample, hardness, total_hard, targets):
```
Comment (sshleifer) [inline fastai/imagito/sample_hardness.py]: I dont get what this does/why called proxy_dist?

File: fastai/imagito/utils.py [hunk 5]
```diff
@@ -39,6 +43,42 @@ def update_batch_size(pg):
         new_pars.append(new_p)
     return new_pars
 
+def filter_classes(image_list, classes=None):
```
Comment (sshleifer) [inline fastai/imagito/utils.py]: classes.py?

File: fastai/basic_data.py [hunk 6]
```diff
@@ -116,7 +116,7 @@ def create(cls, train_ds:Dataset, valid_ds:Dataset, test_ds:Optional[Dataset]=No
         datasets = cls._init_ds(train_ds, valid_ds, test_ds)
         val_bs = ifnone(val_bs, bs)
         dls = [DataLoader(d, b, shuffle=s, drop_last=s, num_workers=num_workers, **dl_kwargs) for d,b,s in
-               zip(datasets, (bs,val_bs,val_bs,val_bs), (True,False,False,False)) if d is not None]
+               zip(datasets, (bs,val_bs,val_bs,val_bs), (False,False,False,False)) if d is not None]
```
Comment (sshleifer) [inline fastai/basic_data.py]: I think we want to keep shuffling train in other experiments, so we should ideally pass this down.

File: fastai/imagito/sample_hardness.py (hunk 4 above)
Comment (ecprokop) [reply fastai/imagito/sample_hardness.py]: `dist` meaning distribution. Just shared logic between sample_loss_based and sample_correctness_based for calculating/rounding total num examples in proxy as well as number of easy/hard examples.

Also warns if number of "hard" examples you wanted isn't available, so you know what you're actually training on.

It wasn't as clean of a decomposition as I thought it'd be at first (`total_hard` can't be calculated the same way for both approaches so I passed it in just to get it working) but it works

File: fastai/basic_data.py (hunk 6 above)
Comment (ecprokop) [reply fastai/basic_data.py]: Ya makes sense, will make the change

File: fastai/imagito/sample_hardness.py (hunk 3 above)
Comment (ecprokop) [reply fastai/imagito/sample_hardness.py]: appreciate the style review for my knowledge and happy to make this and other changes but let's keep in mind we aren't building a production system here; I am rapidly shifting towards "just get shit done" mode for both my projects, i.e. who cares about cosmetics if it works and is reasonably understandable :D

File: fastai/imagito/sample_hardness.py (hunk 2 above)
Comment (ecprokop) [reply fastai/imagito/sample_hardness.py:48]: Seems like it unless you know of a better way?

```
     i = 0
    def mask_filter(img):
        keep = bool(mask[i])
        i += 1
        return keep
```

```
  File "sample_hardness.py", line 47, in mask_filter
    keep = bool(mask[i])
UnboundLocalError: local variable 'i' referenced before assignment
```

File: fastai/imagito/sample_hardness.py (hunk 1 above)
Comment (ecprokop) [reply fastai/imagito/sample_hardness.py:27]: I don't think so – that will bound idx between `[0, n)`

We want n random idx between `[0, t.size())`

File: fastai/imagito/utils.py (hunk 5 above)
Comment (ecprokop) [reply fastai/imagito/utils.py]: will move!

File: fastai/imagito/sample_hardness.py (hunk 4 above)
Comment (sshleifer) [reply fastai/imagito/sample_hardness.py]: cool dist can be also be distance

File: fastai/imagito/sample_hardness.py (hunk 3 above)
Comment (sshleifer) [reply fastai/imagito/sample_hardness.py]: of course :) I just type what I see doesn't mean you need to do anything!

File: fastai/imagito/sample_hardness.py (hunk 4 above)
Comment (ecprokop) [reply fastai/imagito/sample_hardness.py]: oops, that is indeed confusing! changed to `proxy_distr`

File: fastai/imagito/sample_hardness.py (hunk 3 above)
Comment (ecprokop) [reply fastai/imagito/sample_hardness.py]: sg, dank je !

**Pull request merged**